blueprints/api.guidelines.md

```diff
@@ -0,0 +1,549 @@
+property, and because of that there must be a *very convincing* argument if and
+when that is the case.
+
+As a special case, endpoints that report statistics including minimums,
+maximums and arithmetic means of data sets MUST use the property names `min`,
+`max`, and `mean`, respectively, to express those concepts. These SHOULD be
+properties of `response` directly whenever that makes sense.
+
+#### Alerts
+Alerts should be presented as an array containing objects which each conform to
+the object definition laid out by
+[the ATC library's Alert structure](https://godoc.org/github.com/apache/trafficcontrol/lib/go-tc#Alert).
+The allowable `level`s of an Alert are:
+
+- `error` - This level MUST be used to indicate conditions that caused a request
+to fail. Because of this, this level MUST NOT appear in the `alerts` array of
+responses with any HTTP response code less than 400. Details of server workings
+and/or failing components MUST NOT be exposed in this message, which should
+otherwise be as descriptive as possible.
+- `info` - This level SHOULD be used to convey supplementary information to a
+user that is not directly the result of their request. This SHOULD NOT carry
+information indicating whether or not the request succeeded and why/why not, as
+that is best left to the `error` and `success` levels.
+- `success` - This level MUST be used to convey success messages to the client.
+In general, it is expected that the message will be directly displayed to th
+user by the client, and therefore can be used to add human-friendly details
+about a request beyond the response payload. This level MUST NOT appear in the
+`alerts` array of responses with an HTTP response code that is not between 200
+and 399 (inclusive).
+- `warning` - This level is used to warn clients of potentially dangerous
+conditions when said conditions have not caused a request to fail. The best
+example of this is deprecation warnings, which should appear on all API routes
+that have been deprecated.
+
+#### Summary
+The `summary` object is used to provide summary statistics about object
+collections. In general the use of `summary` is left to be defined by API
+endpoints (subject to some restrictions). However, its use is not appropriate in
+cases where the user is specifically requesting summary statistics, but should
+rather be used to provide supporting information - pre-calculated - about a set
+of objects or data that the client *has* requested.
+
+Endpoints MUST use the following, reserved properties of `summary` for their
+described purposes (when use of `summary` is appropriate) rather than defining
+new `summary` or `response` properties to suit the same purpose:
+
+- `count` - Count contains an unsigned integer that defines the total number of
+results that could possibly be returned given the non-pagination query
+parameters supplied by the client.
+
+### HTTP Request Methods
+[RFC 2616 Section 9](https://tools.ietf.org/html/rfc2616#section-9) defines the
```

Comment:
RFC 2616 is obsolete, should be RFC 7231 Section 4: https://tools.ietf.org/html/rfc7231#section-4